homeassistant/components/rainbird/calendar.py

```diff
@@ -0,0 +1,159 @@
+        self, hass: HomeAssistant, start_date: datetime, end_date: datetime
+    ) -> list[CalendarEvent]:
+        """Get all events in a specific time frame."""
+        schedule = self.coordinator.data
+        if not schedule:
+            raise HomeAssistantError(
+                "Unable to get events: No data from controller yet"
+            )
+        cursor = schedule.timeline_tz(dt_util.DEFAULT_TIME_ZONE).overlapping(
+            dt_util.as_local(start_date), dt_util.as_local(end_date)
+        )
+        return [
+            CalendarEvent(
+                summary=program_event.program_id.name,
+                start=dt_util.as_local(program_event.start),
+                end=dt_util.as_local(program_event.end),
+                rrule=program_event.rrule_str,
+            )
+            for program_event in cursor
+        ]
+
+    def _apply_coordinator_update(self) -> None:
+        """Copy state from the coordinator to this entity."""
+        schedule = self.coordinator.data
+        if not schedule:
+            _LOGGER.debug("No schedule")
+            self._event = None
+            return
+        cursor = schedule.timeline_tz(dt_util.DEFAULT_TIME_ZONE).active_after(
+            dt_util.now()
+        )
+        program_event = next(cursor, None)
+        if not program_event:
+            _LOGGER.debug("No program event")
+            self._event = None
+            return
+        _LOGGER.debug("Event=%s", program_event.start)
+        self._event = CalendarEvent(
+            summary=program_event.program_id.name,
+            start=dt_util.as_local(program_event.start),
+            end=dt_util.as_local(program_event.end),
+            rrule=program_event.rrule_str,
+        )
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._apply_coordinator_update()
+        super()._handle_coordinator_update()
+
+    async def async_update(self) -> None:
+        """Disable update behavior.
+
+        This disables the parent class behavior that asks the update coordinator to refresh.
+        The coordinator itself updates at a slower pace (actually sending device RPCs) and
+        this entity also updates itself to advance the clock and check if the next upcoming
+        event is active. We don't call the update cordinator here, but do allow the calendar
+        event state updates to happen that evaluate `event`.
+        """
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to hass."""
+        await super().async_added_to_hass()
+
+        # We do not ask for an update with async_add_entities()
+        # because it will update disabled entities. This is started as a
+        # task to let it sync in the background without blocking startup
+        async def refresh() -> None:
+            await self.coordinator.async_request_refresh()
+            self._apply_coordinator_update()
+
```

Comment:
Can this be a bound method instead? It doesn't look like it needs to be recreated each time

Reply:
Let me know if this is not what you had in mind.

Reply:
Thats better (I'd call it `_async_refresh` though).  I don't think you would ever have enough of these that it makes much difference here but we have problems with entities that generate a new callable when added to hass when we get to hundreds of them since it means we end up with hundreds of copies of the function in memory instead of just the reference to the bound method.

Reply:
This has been removed given the other calendar entity simplifications.